### DISCUSSION TITLE
v8 and CRD schema generation

### DISCUSSION THREAD

**Discussion by slacki123** [dotnet/dotnet-operator-sdk#696] [Q&A] [answered]
Hi @buehler, I've been following the recent updates to kubeops and I'm quite impressed with the extensive changes introduced in version 8. I'm particularly interested in understanding more about the updates related to CRD schema generation.

Would you be able to guide me towards the relevant documentation or provide some insight into how these changes might alter current functionalities? Any assistance in this regard would be greatly appreciated.

**Comment by buehler** [comment]
Hey @slacki123 

First: Thank you 😊 

What do you mean by "alter current functionalities" ? I tried to stay true to all the features that were already present. So in terms of functionality, there _should_ be no changes. The tests are the same. The changes that I made mainly split up the whole thing.

To (very generously) summarize: There are more packages (separation of concern), you can make a simple, non webhook operator without ASP.net now, the webhooks still exist, but only with ASP.net, and the CLI is now a dotnet tool instead of baked into the whole thing.

The CLI does load any assembly and parses the type information to generate CRDs and stuff.

The result should be the same, but it is not ran from the "KubeOps" package as a whole, but the CLI loads assemblies and parses type info from there.

Does this answer your question?
Cheeres

**Reply by slacki123** [reply]
Ah I see. So the algorithm to recursively generate a schema from objects should work the same I understand. I maybe looked at it wrong previously and assumed there has been some changes. But that's fair enough, my question is answered. Thanks!

**Reply by buehler** [reply, marked as answer]
Yes, the algorithm is the same. Just in another place :-)

However, it was just massively simplified by @mauleb. It still is recursive, but not as shakey anymore (since some types and compile steps were involved in the CLI)
